File: android/src/main/java/com/fitnessgeolocation/GeoMath.kt
package com.fitnessgeolocation

/** Shared geo primitives — ray-cast O(n), bbox reject O(1). */
object GeoMath {
  data class Point(val lat: Double, val lng: Double)

  /** Ray-casting point-in-polygon (lat/lng vertices). */
  fun pointInPolygon(lat: Double, lng: Double, vertices: List<Point>): Boolean {
    val n = vertices.size
    if (n < 3) return false
    var inside = false
    var j = n - 1
    for (i in 0 until n) {
      val yi = vertices[i].lat; val xi = vertices[i].lng
      val yj = vertices[j].lat; val xj = vertices[j].lng
      if ((yi > lat) != (yj > lat) &&
        lng < (xj - xi) * (lat - yi) / (yj - yi + 1e-12) + xi
      ) inside = !inside
      j = i
    }
    return inside
  }

  fun boundingBox(vertices: List<Point>): DoubleArray? {
    if (vertices.isEmpty()) return null
    var minLat = vertices[0].lat; var maxLat = minLat
    var minLng = vertices[0].lng; var maxLng = minLng
    for (v in vertices.drop(1)) {
      minLat = minOf(minLat, v.lat); maxLat = maxOf(maxLat, v.lat)
      minLng = minOf(minLng, v.lng); maxLng = maxOf(maxLng, v.lng)
    }
    return doubleArrayOf(minLat, maxLat, minLng, maxLng)
  }

  fun inBoundingBox(lat: Double, lng: Double, box: DoubleArray): Boolean =
    lat in box[0]..box[1] && lng in box[2]..box[3]

  fun parseVertices(raw: List<Map<String, Any?>>): List<Point> =
    raw.mapNotNull { v ->
      val lat = (v["latitude"] as? Number)?.toDouble() ?: return@mapNotNull null
      val lng = (v["longitude"] as? Number)?.toDouble() ?: return@mapNotNull null
      Point(lat, lng)
    }

  /**
   * Haversine distance in meters.
   * Complexity: O(1)
   */
  fun haversineMeters(lat1: Double, lng1: Double, lat2: Double, lng2: Double): Double {
    val R = 6371000.0
    val dLat = Math.toRadians(lat2 - lat1)
    val dLon = Math.toRadians(lng2 - lng1)
    val a = Math.sin(dLat / 2) * Math.sin(dLat / 2) +
      Math.cos(Math.toRadians(lat1)) * Math.cos(Math.toRadians(lat2)) *
      Math.sin(dLon / 2) * Math.sin(dLon / 2)
    val c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a))
    return R * c
  }
}

File: ios/FitnessGeolocation/GeoMath.swift
import Foundation

/// Shared geo primitives — O(n) ray-cast, O(1) bbox reject. Used by polygon geofences.
enum GeoMath {
  struct Point { let lat: Double; let lng: Double }

  /// Ray-casting point-in-polygon. Vertices as (lat, lng) pairs, closed or open ring.
  static func pointInPolygon(lat: Double, lng: Double, vertices: [Point]) -> Bool {
    let n = vertices.count
    guard n >= 3 else { return false }
    var inside = false
    var j = n - 1
    for i in 0..<n {
      let yi = vertices[i].lat, xi = vertices[i].lng
      let yj = vertices[j].lat, xj = vertices[j].lng
      if ((yi > lat) != (yj > lat)) &&
          (lng < (xj - xi) * (lat - yi) / (yj - yi + 1e-12) + xi) {
        inside.toggle()
      }
      j = i
    }
    return inside
  }

  static func boundingBox(_ vertices: [Point]) -> (minLat: Double, maxLat: Double, minLng: Double, maxLng: Double)? {
    guard !vertices.isEmpty else { return nil }
    var minLat = vertices[0].lat, maxLat = minLat
    var minLng = vertices[0].lng, maxLng = minLng
    for v in vertices.dropFirst() {
      minLat = min(minLat, v.lat); maxLat = max(maxLat, v.lat)
      minLng = min(minLng, v.lng); maxLng = max(maxLng, v.lng)
    }
    return (minLat, maxLat, minLng, maxLng)
  }

  static func inBoundingBox(lat: Double, lng: Double, box: (minLat: Double, maxLat: Double, minLng: Double, maxLng: Double)) -> Bool {
    lat >= box.minLat && lat <= box.maxLat && lng >= box.minLng && lng <= box.maxLng
  }

  static func parseVertices(_ raw: [[String: Any]]) -> [Point] {
    raw.compactMap { v in
      guard let lat = v["latitude"] as? Double, let lng = v["longitude"] as? Double else { return nil }
      return Point(lat: lat, lng: lng)
    }
  }
}
